File: 740-delete-and-earn/740-delete-and-earn.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        points = defaultdict(int)
        # Precompute how many points we gain from taking an element
        for num in nums:
            points[num] += num
            
        elements = sorted(points.keys())
        two_back = 0
        one_back = points[elements[0]]
        
        for i in range(1, len(elements)):
            current_element = elements[i]
            if current_element == elements[i - 1] + 1:
                # The 2 elements are adjacent, cannot take both - apply normal recurrence
                two_back, one_back = one_back, max(one_back, two_back + points[current_element])
            else:
                # Otherwise, we don't need to worry about adjacent deletions
                two_back, one_back = one_back, one_back + points[current_element]

        return one_back

File: 740-delete-and-earn/740-delete-and-earn.cpp
class Solution {
public:
    map<int,int> mp;
    int solve(int i,int n,vector<int>&arr,vector<int>&dp){
        if(i>=10001) return 0;
        if(dp[i]!=-1) return dp[i];
        
        return dp[i]=max(mp[i]*i+solve(i+2,n,arr,dp),solve(i+1,n,arr,dp));
    }
    
    int deleteAndEarn(vector<int>& nums) {
        for(auto x:nums){
            mp[x]++;
        }
        vector<int> arr;
        for(auto k:mp){
            arr.push_back(k.first);
        }
        int n = arr.size();
        vector<int> dp(10001,-1);
        return solve(0,n,arr,dp);
    }
};